core/src/impl/Kokkos_Half_MathematicalFunctions.hpp

```diff
@@ -0,0 +1,214 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#ifndef KOKKOS_HALF_MATHEMATICAL_FUNCTIONS_HPP_
+#define KOKKOS_HALF_MATHEMATICAL_FUNCTIONS_HPP_
+
+#include <Kokkos_MathematicalFunctions.hpp>  // For the float overloads
+
+namespace Kokkos {
+////////////// BEGIN HALF_T (float16) MATH FNS ////////////
+#if defined(KOKKOS_HALF_T_IS_FLOAT) && !KOKKOS_HALF_T_IS_FLOAT
+#define KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(FUNC)          \
+  KOKKOS_INLINE_FUNCTION Kokkos::Experimental::half_t FUNC( \
+      Kokkos::Experimental::half_t x) {                     \
+    return static_cast<Kokkos::Experimental::half_t>(       \
+        Kokkos::FUNC(static_cast<float>(x)));               \
+  }
+
```

Comment:
Is there any reason not to do half_t and bhalf_t in one go if both are available? Are there any math functions which don't work in both cases?

```c++
#define KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF_TYPE(FUNC, TYPE)          \
  KOKKOS_INLINE_FUNCTION TYPE FUNC( TYPE x)                          \
    { return static_cast<TYPE>(  Kokkos::FUNC(static_cast<float>(x)));  }
    
#if defined(KOKKOS_HALF_T_IS_FLOAT) && !KOKKOS_HALF_T_IS_FLOAT
  #if defined(KOKKOS_BHALF_T_IS_FLOAT) && !KOKKOS_BHALF_T_IS_FLOAT
    #define KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(FUNC) \
            KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF_TYPE(FUNC, Kokkos::Experimental::half_t) \
            KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF_TYPE(FUNC, Kokkos::Experimental::bhalf_t)
  #else
    #define KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(FUNC) \
         KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF_TYPE(FUNC, Kokkos::Experimental::half_t)
  #endif
#endif
```

This way we won't overlook any instantiation.